daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala

```diff
@@ -1330,7 +1345,8 @@ object Main {
                   case s: scala.util.control.ControlThrowable => throw s
                   case u: UnsuppressableException => throw u
                   case e: TDMLTestNotCompatibleException => {
-                    println("[Skipped] %s (Not compatible implementation.)".format(name))
+                    println("[Skipped] %s (not compatible with implementation %s)"
+                      .format(name, e.implementation.getOrElse("<none>")))
```

Comment:
We must have an implementation here, right? How could a test not be compatible with the implementation yet the exception doesn't know what implementation isn't compatible?

Digging a bit, it looks like we just carry around the implementation in the TDMLRunner as an `Option`. Which doesn't really make sense. We must always know what implementation is being used. In fact, often times we just call `impl.get`, so there's definitely an assumption that it's defined.

This is an existing problem so I don't suggest you fix it here, but maybe we should open a new bug to change references to the implementation in the TDMLRunner to not be an Option? That might be a good beginner bug for someone.

Reply:
Yeah, it seems reasonable for `TDMLException` to have an `Option` implementation--some exceptions definitely happen before we know what the implementation is.

But I don't think it shouldn't be an `Option` in the `TDMLTestNotCompatibleImplementation` exception. Ideally that class should be changed to accept a `String` and it creates a `Some(implementation)` when passed to the TDMLExceptionImpl. I'm not sure if that's worth that change, but you're right that TDMLException is probably the reason why it's an an Option.